store/cachemulti/store.go

```diff
@@ -0,0 +1,135 @@
+package cachemulti
+
+import (
+	"io"
+
+	dbm "github.com/tendermint/tendermint/libs/db"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/dbadapter"
+	"github.com/cosmos/cosmos-sdk/store/types"
+)
+
+//----------------------------------------
+// Store
+
+// Store holds many cache-wrapped stores.
+// Implements MultiStore.
+// NOTE: a Store (and MultiStores in general) should never expose the
+// keys for the substores.
+type Store struct {
+	db     types.CacheKVStore
+	stores map[types.StoreKey]types.CacheWrap
+	keys   map[string]types.StoreKey
+
+	traceWriter  io.Writer
+	traceContext types.TraceContext
+}
+
+var _ types.CacheMultiStore = Store{}
+
+func NewFromKVStore(
+	store types.KVStore,
+	stores map[types.StoreKey]types.CacheWrapper, keys map[string]types.StoreKey,
+	traceWriter io.Writer, traceContext types.TraceContext,
+) Store {
+	cms := Store{
+		db:           cachekv.NewStore(store),
+		stores:       make(map[types.StoreKey]types.CacheWrap, len(stores)),
+		keys:         keys,
+		traceWriter:  traceWriter,
+		traceContext: traceContext,
+	}
+
+	for key, store := range stores {
+		if cms.TracingEnabled() {
+			cms.stores[key] = store.CacheWrapWithTrace(cms.traceWriter, cms.traceContext)
+		} else {
+			cms.stores[key] = store.CacheWrap()
+		}
+	}
+
+	return cms
+}
+
+func NewStore(
+	db dbm.DB,
+	stores map[types.StoreKey]types.CacheWrapper, keys map[string]types.StoreKey,
+	traceWriter io.Writer, traceContext types.TraceContext,
+) Store {
+	return NewFromKVStore(dbadapter.Store{db}, stores, keys, traceWriter, traceContext)
+}
+
+func newCacheMultiStoreFromCMS(cms Store) Store {
+	stores := make(map[types.StoreKey]types.CacheWrapper)
+	for k, v := range cms.stores {
+		stores[k] = v
+	}
+	return NewFromKVStore(cms.db, stores, nil, cms.traceWriter, cms.traceContext)
+}
+
+// SetTracer sets the tracer for the MultiStore that the underlying
+// stores will utilize to trace operations. A MultiStore is returned.
+func (cms Store) SetTracer(w io.Writer) types.MultiStore {
```

Comment:
I think `WithTracer` makes more sense here since `SetXX` usually does not return anything.

Reply:
It is also weird to use `WithXXX` here because normally `WithXXX` does not modify the original value, for example `Context.WithXXX`. `SetTracer` and `SetTracingContext` modifies original `MultiStore`, so it would be nice to name the functions depending on the mutability I guess(since they are not shown in the function signature, unlike whether the function returns a value or not). Also see https://github.com/cosmos/cosmos-sdk/issues/2824